### PR TITLE
! operator

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

Please note, the first commit is PR #21309, the subsequent commit is support for the not-null postfix `!` operator.

**Comment by AlekseyTs** [comment]
It looks like no tests were executed for this PR.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 1]
```diff
@@ -2002,6 +2002,20 @@ private BoundExpression BindIncrementOperator(CSharpSyntaxNode node, ExpressionS
                 hasErrors);
         }
 
+        private BoundExpression BindNotNullExpression(PostfixUnaryExpressionSyntax node, DiagnosticBag diagnostics)
+        {
+            var expr = BindValue(node.Operand, diagnostics, BindValueKind.RValue);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: >BindValueKind.RValue [](start = 60, length = 20)

Is this always an RValue though? Can I use ```!``` on a ```ref``` parameter, or in another "ref" context?

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 2]
```diff
@@ -2002,6 +2002,20 @@ private BoundExpression BindIncrementOperator(CSharpSyntaxNode node, ExpressionS
...
+            var type = expr.Type;
+            if ((object)type != null &&
+                !type.IsErrorType() &&
+                !type.IsReferenceType &&
+                Compilation.IsFeatureEnabled(MessageID.IDS_FeatureStaticNullChecking))
+            {
+                Error(diagnostics, ErrorCode.ERR_NotNullableOperatorNotReferenceType, node);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: >Error(diagnostics, ErrorCode.ERR_NotNullableOperatorNotReferenceType, node); [](start = 16, length = 76)

I would expect this error to be unconditional.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: `!` can be used in `ref` contexts (see `NotNullOperator_RefOut` test) although those cases are not handled yet.

**Comment by cston** [comment]
@dotnet-bot retest this please

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: Then I would say we shouldn't be binding the underlying expression as an RValue.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: We probably should bind in the same manner BindParenthesizedExpression binds underlying expression and then adjust CheckValue (and the like) to understand the "pass-through" nature of BoundNotNullExpression.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: I mean with respect to whether the feature is enabled or not.

File: src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs [hunk 3]
```diff
@@ -317,17 +317,7 @@ public char PeekChar(int delta)
...
-                && !MoreChars())
-            {
-                ch = InvalidCharacter;
-            }
-            else
-            {
-                // N.B. MoreChars may update the offset.
-                ch = _characterWindow[_offset];
-            }
+            char ch = PeekChar();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs]: >char ch = PeekChar(); [](start = 12, length = 21)

This change looks unrelated to the feature. Consider moving it to a separate PR and merging it to master.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs]: #21626

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 4]
```diff
@@ -2027,6 +2030,8 @@ internal bool CheckValueKind(BoundExpression expr, BindValueKind valueKind, Diag
                 case BoundKind.DynamicMemberAccess:
                 case BoundKind.DynamicIndexerAccess:
                     return CheckDynamicValueKind(expr, valueKind, diagnostics);
+                case BoundKind.NotNullExpression:
+                    return CheckValueKind(((BoundNotNullExpression)expr).Expression, valueKind, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >return CheckValueKind(((BoundNotNullExpression)expr).Expression, valueKind, diagnostics); [](start = 20, length = 89)

I believe ```private BoundExpression CheckValue(BoundExpression expr, BindValueKind valueKind, DiagnosticBag diagnostics)``` should also be changed to dig through the ```BoundNotNullExpression``` and preserve any transformations made to the underlying expression in the process.

**Comment by AlekseyTs** [comment]
>         t! = s; // parsed as t != s;

I would expect some errors for this code.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:15083 in a047ef7. [](commit_id = a047ef75a2ff41a762e0c9e435d534f6150e847a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
It looks like there are some legitimate test failures.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 5]
```diff
@@ -238,7 +238,12 @@
     <Field Name="MethodOpt" Type="MethodSymbol" Null="allow"/>
     <Field Name="ResultKind" PropertyOverrides="true" Type="LookupResultKind"/>
   </Node>
-  
+
+  <Node Name="BoundNotNullExpression" Base="BoundExpression">
```
Comment (DustinCampbell) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Sigh... I was really hoping for `DammitExpression` :smile:

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 4 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: Good catch. Updated `CheckValue`.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 5 above)
Comment (alrz) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: How about
```xml 
  <Node Name="PostfixUnaryExpressionSyntax" Base="ExpressionSyntax">
    <Kind Name="ForceValueExpression"/>
    <Field Name="OperatorToken" Type="SyntaxToken">
      <Kind Name="ExclamationToken"/>
    </Field>
    <Field Name="Operand" Type="ExpressionSyntax"/>
  </Node>
```

Comment (alrz) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: How about `BoundForceValueExpression`?

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review. Feel free to ignore the first commit which is #21309. Thanks.

**Comment by AlekseyTs** [comment]
https://ci.dot.net/job/dotnet_roslyn/job/features_NullableReferenceTypes/job/windows_release_vs-integration_prtest/8/

CLR test failure (only matches the first, see "Latest Test Result" link for the rest)Roslyn.VisualStudio.IntegrationTests.VisualBasic.BasicExpressionEvaluator.EvaluatePrimitiveValues [FAIL]Indication 1
--
  | There are errors when processing test resultsThere was an error processing the xunit test results: [xUnit] [ERROR] - No test reports found for the metric 'xUnit.Net' with the resolved pattern '**/xUnitResults/*.xml'. Configuration error?.Indication 2

**Comment by AlekseyTs** [comment]
Test windows_release_vs-integration_prtest please

**Comment by AlekseyTs** [comment]
>         t! = s;

Consider also testing ```(t!) = s;```

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:15084 in fbaceff. [](commit_id = fbaceffba46f80be9c3f14dbb47a14c014838698, deletion_comment = False)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 6]
```diff
@@ -238,7 +238,12 @@
...
+    <Field Name="Type" Type="TypeSymbol" Override="true" Null="allow"/>
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: I believe the principal intended semantics are to suppress a warning, not make the compiler think the value is not null. At least, that's the latest view. Perhaps BoundSuppressNullableWarningExpression?

File: src/Compilers/Core/Portable/Operations/INotNullExpression.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+namespace Microsoft.CodeAnalysis.Semantics
+{
+    /// <summary>
+    /// Represents an expression with a type that should be treated as not null.
+    /// </summary>
+    /// <remarks>
+    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
+    /// change it in the future.
+    /// </remarks>
+    public interface INotNullExpression : IOperation
```
Comment (gafter) [inline src/Compilers/Core/Portable/Operations/INotNullExpression.cs]: >INotNullExpression [](start = 21, length = 18)

ISuppressNullableWarningExpression?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -9333,6 +9333,12 @@ private ExpressionSyntax ParsePostFixExpression(ExpressionSyntax expr)
                         }
 
                         goto default;
+
+                    case SyntaxKind.ExclamationToken:
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >case [](start = 20, length = 4)

I gather the syntax is
``` antlr
primary-no-array-creation-expression
    : exclamation-expression
    ;
exclamation-expression
    : primary-expression '!'
    ;
```

In which case it must be a *semantic* error when you write `e!!`.

**Comment by gafter** [comment]
>         }

I think we decided that `!!` is supposed to be an error.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:15012 in fbaceff. [](commit_id = fbaceffba46f80be9c3f14dbb47a14c014838698, deletion_comment = False)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 6 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: I'll rename to `BoundSuppressNullableWarningExpression".

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 8 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I've added a commented out test for `!!` and will add the binding error in a subsequent PR.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 6 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: FWIW if this could be used on `Nullable<T>` it's not just suppressing warnings. it force-unwrap the value.

Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: We can rename the node again if the scope or meaning change.

File: src/Compilers/CSharp/Portable/BoundTree/Formatting.cs [hunk 9]
```diff
@@ -38,6 +38,14 @@ public override object Display
         }
     }
 
+    internal sealed partial class BoundSuppressNullableWarningExpression
+    {
+        public override object Display
+        {
+            get { return Expression.Display + "!"; }
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/Formatting.cs]: Please `Expression.Type ?? Expression.Display + "!"`

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 10]
```diff
@@ -5234,6 +5234,9 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_ExplicitNullableAttribute" xml:space="preserve">
     <value>Explicit application of 'System.Runtime.CompilerServices.NullableAttribute' is not allowed.</value>
   </data>
+  <data name="ERR_NotNullableOperatorNotReferenceType" xml:space="preserve">
+    <value>The ! operator can only be applied to reference types.</value>
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Probably not right, e.g. `Box<string?>` could be banged.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs [hunk 11]
```diff
@@ -4021,6 +4006,19 @@ public override BoundNode VisitAsOperator(BoundAsOperator node)
             return result;
         }
 
+        public override BoundNode VisitSuppressNullableWarningExpression(BoundSuppressNullableWarningExpression node)
+        {
+            var result = base.VisitSuppressNullableWarningExpression(node);
+
+            Debug.Assert(!IsConditionalState);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: >Debug [](start = 12, length = 5)

Assert would fail if I write `if (c!) {}`

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs [hunk 12]
```diff
@@ -4021,6 +4006,19 @@ public override BoundNode VisitAsOperator(BoundAsOperator node)
...
+            if (_performStaticNullChecks && this.State.Reachable)
+            {
+                this.State.ResultIsNotNull = true;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: >this [](start = 16, length = 4)

Suggest either making it oblivious or implementing the suppression at the use site.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs [hunk 13]
```diff
@@ -4021,6 +4006,19 @@ public override BoundNode VisitAsOperator(BoundAsOperator node)
             return result;
         }
 
+        public override BoundNode VisitSuppressNullableWarningExpression(BoundSuppressNullableWarningExpression node)
+        {
+            var result = base.VisitSuppressNullableWarningExpression(node);
+
+            Debug.Assert(!IsConditionalState);
+            if (_performStaticNullChecks && this.State.Reachable)
+            {
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: If the state when a variable is either uninitialized or in unreachable code were to be oblivious, I think that gives you the right semantics without having to check reachable state anywhere.

File: src/Compilers/CSharp/Test/Syntax/Parsing/SuppressNullableWarningExpressionParsingTests.cs [hunk 14]
```diff
@@ -0,0 +1,466 @@
...
+        {
+            var source =
+@"class C
+{
+    static void F(object? o)
+    {
+        o = o!==null;
+    }
+}";
+            var tree = UsingTree(source);
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/SuppressNullableWarningExpressionParsingTests.cs]: >var [](start = 12, length = 3)

Please verify errors.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs (hunk 11 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: `IsConditionalState` is always `true` currently because `base.VisitSuppressNullableWarningExpression()` calls `VisitRvalue(node.Expression)`. For an expression such as `(b && F(out var o))!? o : null)`, the condition should not be treated as an rvalue. Added `SuppressNullableWarning_Condition` unit test for now.

File: src/Compilers/CSharp/Portable/BoundTree/Formatting.cs (hunk 9 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/Formatting.cs]: Done.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 10 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/CSharpResources.resx]: Added `SuppressNullableWarning_ValueType2` test.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs (hunk 12 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: Changed to oblivious (`null`).

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs (hunk 13 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: Add `// PROTOTYPE` comment to address later.

File: src/Compilers/CSharp/Test/Syntax/Parsing/SuppressNullableWarningExpressionParsingTests.cs (hunk 14 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Syntax/Parsing/SuppressNullableWarningExpressionParsingTests.cs]: Done.
